// @cord.network/config/lib/esm/ConfigService.js
/**
 * The ConfigService is used for setting up the node address,
 * the logging level as well as storing custom configuration options.
 *
 * @packageDocumentation
 * @module ConfigService
 */
import { LFService, LoggerFactoryOptions, LogGroupRule, LogLevel, getLogControl, } from 'typescript-logging';
import { SDKErrors } from '@cord.network/utils';
const DEFAULT_DEBUG_LEVEL = typeof process !== 'undefined' &&
    process.env?.DEBUG &&
    process.env.DEBUG === 'true'
    ? LogLevel.Debug
    : LogLevel.Error;
/**
 * Changes all existing Loggers of our default Factory with id 0 to the intended Level.
 *
 * @param level The intended LogLevel. LogLevel has a range of 0 to 5.
 * @returns The new set level.
 */
export function modifyLogLevel(level) {
    // eslint-disable-next-line no-nested-ternary
    const actualLevel = level > 0 ? (level > 5 ? 5 : level) : 0;
    getLogControl()
        .getLoggerFactoryControl(0)
        .change({
        group: 'all',
        logLevel: LogLevel[actualLevel],
    });
    return actualLevel;
}
const defaultConfig = {
    logLevel: DEFAULT_DEBUG_LEVEL,
};
let configuration = { ...defaultConfig };
/**
 * Get the value set for a configuration.
 *
 * @param configOpt Key of the configuration.
 * @returns Value for this key.
 */
export function get(configOpt) {
    if (typeof configuration[configOpt] === 'undefined') {
        switch (configOpt) {
            case 'api':
                throw new SDKErrors.BlockchainApiMissingError();
            default:
                throw new Error(`GENERIC NOT CONFIGURED ERROR FOR KEY: "${configOpt}"`);
        }
    }
    return configuration[configOpt];
}
function setLogLevel(logLevel) {
    if (logLevel !== undefined) {
        modifyLogLevel(logLevel);
    }
}
/**
 * Set values for one or multiple configurations.
 *
 * @param opts Object of configurations as key-value pairs.
 */
export function set(opts) {
    configuration = { ...configuration, ...opts };
    setLogLevel(configuration.logLevel);
}
/**
 * Set the value for a configuration to its default (which may be `undefined`).
 *
 * @param key Key identifying the configuration option.
 */
export function unset(key) {
    if (Object.prototype.hasOwnProperty.call(defaultConfig, key)) {
        configuration[key] = defaultConfig[key];
    }
    else {
        delete configuration[key];
    }
}
/**
 * Indicates whether a configuration option is set.
 *
 * @param key Key identifying the configuration option.
 * @returns True if this value is set, false otherwise.
 */
export function isSet(key) {
    return typeof configuration[key] !== 'undefined';
}
// Create options instance and specify 1 LogGroupRule:
// * LogLevel Error on default, env DEBUG = 'true' changes Level to Debug.throws
const options = new LoggerFactoryOptions().addLogGroupRule(new LogGroupRule(new RegExp('.+'), get('logLevel')));
// Create a named logging factory and pass in the options and export the factory.
// Named is since version 0.2.+ (it's recommended for future usage)
export const LoggingFactory = LFService.createNamedLoggerFactory('LoggerFactory', options);
